adr/ADR-15.md

```diff
@@ -0,0 +1,233 @@
+# JetStream Subscribe Workflow
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-08-11|
+|Author  |@kozlovic|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client|
+
+## Context
+
+This document attempts to describe the workflow of a JetStream subscription in the library, from the creation, the behavior at runtime, to the deletion of a subscription.
+
+## Design
+
+### Creation
+
+The library should have API(s) that allow(s) creation of a JetStream subscription. Depending on the language, there may be various type of subscriptions (like Channel based in Go), or with various options such as Queue subscriptions, etc... There can be different APIs or one with a set of options.
+
+#### Configuration
+
+When creating a subscription, the user can provide several things:
+
+- A subject that is in some cases required (see "Stream name" section below)
+- An optional stream name
+- An optional consumer name
+- An optional JS consumer configuration (see below), if the library is expected to create the JS consumer
+- A Queue name if this subscription is meant to be a queue subscription
+- Indication if this is a "Pull" subscription or not
+- An optional way of indicating the the given subject will be used to create the internal subscription, by-passing any lookup/consumer creation. For the purpose of this document, let's call this option `SubjectIsDelivery`.
+
+Some misconfiguration should be checked by the subscribe API and return an error outright.
+
```

Comment:
what about `Handlers for receiving messages asynchronously, whatever that means in the library language. `

Reply:
Ok, can add that, but you have sync subscriptions too, and what if there is a library that does not have the concept of async, I don't know. But sure, I can put that on.